[RISCV] Introduce pass to promote double constants to a global array

asb opened the pull request:
As discussed in #153402, we have inefficiences in handling constant pool access that are difficult to address. Using an IR pass to promote double constants to a global allows a higher degree of control of code generation for these accesses, resulting in improved performance on benchmarks that might otherwise have high register pressure due to accessing constant pool values separately rather than via a common base.

Directly promoting double constants to separate global values and relying on the global merger to do a sensible thing would be one potential avenue to explore, but it is _not_ done in this version of the patch because:
* The global merger pass needs fixes. For instance it claims to be a function pass, yet all of the work is done in initialisation. This means that attempts by backends to schedule it after a given module pass don't actually work as expected.
* The heuristics used can impact codegen unexpectedly, so I worry that tweaking it to get the behaviour desired for promoted constants may lead to other issues. This may be completely tractable though.

Now that #159352 has landed, the impact on terms if dynamically executed instructions is slightly smaller (as we are starting from a better baseline), but still worthwhile in lbm and nab from SPEC. Results below are for rva22u64:

```
Benchmark                  Baseline         This PR   Diff (%)
============================================================
============================================================
500.perlbench_r         180668945687    180666122417     -0.00%
502.gcc_r               221274522161    221277565086      0.00%
505.mcf_r               134656204033    134656204066      0.00%
508.namd_r              217646645332    216699783858     -0.44%
510.parest_r            291731988950    291916190776      0.06%
511.povray_r             30983594866     31107718817      0.40%
519.lbm_r                91217999812     87405361395     -4.18%
520.omnetpp_r           137699867177    137674535853     -0.02%
523.xalancbmk_r         284730719514    284734023366      0.00%
525.x264_r              379107521547    379100250568     -0.00%
526.blender_r           659391437610    659447919505      0.01%
531.deepsjeng_r         350038121654    350038121656      0.00%
538.imagick_r           238568674979    238560772162     -0.00%
541.leela_r             405660852855    405654701346     -0.00%
544.nab_r               398215801848    391352111262     -1.72%
557.xz_r                129832192047    129832192055      0.00%

```

---
Notes for reviewers:
* As discussed at the sync-up meeting, the suggestion is to try to land an incremental improvement to the status quo even if there is more work to be done around the general issue of constant pool handling. We can discuss here if that is actually the best next step or not, but I just wanted to clarify that's why this is being posted with a somewhat narrow scope.
* I've disabled transformations both for RV32 and on systems without D as bath cases saw some regressions.

asb committed:
[RISCV] Introduce pass to promote double constants to a global array

As discussed in #153402, we have inefficiences in handling constant pool
access that are difficult to address. Using an IR pass to promote
double constants to a global allows a higher degree of control of code
generation for these accesses, resulting in improved performance on
benchmarks that might otherwise have high register pressure due to
accessing constant pool values separately rather than via a common base.

Directly promoting double constants to separate global values and
relying on the global merger to do a sensible thing would be one
potential avenue to explore, but it is _not_ done in this version of the
patch because:
* The global merger pass needs fixes. For instance it claims to be a
  function pass, yet all of the work is done in initialisation. This
  means that attempts by backends to schedule it after a given module
  pass don't actually work as expected.
* The heuristics used can impact codegen unexpectedly, so I worry that
  tweaking it to get the behaviour desired for promoted constants may
  lead to other issues. This may be completely tractable though.

Now that #159352 has landed, the impact on terms if dynamically executed
instructions is slightly smaller (as we are starting from a better
baseline), but still worthwhile in lbm and nab from SPEC. Results below
are for rva22u64:

```
Benchmark                  Baseline         This PR   Diff (%)
============================================================
500.perlbench_r         180667466583    180667466661      0.00%
502.gcc_r               221281439537    221277561043     -0.00%
505.mcf_r               134656203905    134656204017      0.00%
508.namd_r              217646645213    217616374477     -0.01%
510.parest_r            291730242760    291917069933      0.06%
511.povray_r             30982459833     31101871667      0.39%
519.lbm_r                91217999812     89029313608     -2.40%
520.omnetpp_r           137705551722    138044390554      0.25%
523.xalancbmk_r         284733326286    284728940808     -0.00%
525.x264_r              379107521545    379100249676     -0.00%
526.blender_r           659391437704    659446918261      0.01%
531.deepsjeng_r         350038121655    350038121654     -0.00%
538.imagick_r           238568679271    238560769465     -0.00%
541.leela_r             405654701351    405660852862      0.00%
544.nab_r               398215801713    391380811065     -1.72%
557.xz_r                129832192046    129832192047      0.00%
```

asb committed:
formatting

lukel97 commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,212 @@
+  // offset within the newly created global array. Also track uses that will
+  // be replaced later.
+  DenseMap<ConstantFP *, unsigned> ConstantMap;
+  SmallVector<Constant *, 16> ConstantVector;
+  DenseMap<ConstantFP *, SmallVector<Use *, 8>> UsesInFunc;
```
Are the keys to UsesInFunc the same as ConstantVector? If so could you use a MapVector instead for deterministic insertion order?

asb replied:
Done - this was a nice simplification so thank you.

preames commented on llvm/test/CodeGen/RISCV/double-imm.ll:
```diff
@@ -17,8 +17,8 @@ define double @double_imm() nounwind {
 ; CHECK64D:       # %bb.0:
-; CHECK64D-NEXT:    lui a0, %hi(.LCPI0_0)
-; CHECK64D-NEXT:    fld fa0, %lo(.LCPI0_0)(a0)
+; CHECK64D-NEXT:    lui a0, %hi(.promoted_doubles.double_imm)
```
Hm, looks like maybe we shouldn't apply the transform unless there's more than one double in the function?

asb replied:
Agreed, added.

preames reviewed (commented):
Can you add some tests specifically for this transform?  IR to IR is fine.

arsenm reviewed (commented):
Should have new IR test. Not sure really why this needs to be a new IR pass and not just a lowering behavior in codegen?

asb committed:
Address review comments

asb commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  // Collect uses from PHI nodes after other uses, because when transforming
+  // the function, we handle PHI uses afterwards.
```
Yes it's the awkward issue related to choosing a different insertion point when the use is a PHI or for a "normal" use.

But taking a step back, I feel I've failed to simplify this as much as possible. I'd started like AArch64ConstantPromotion with doing minimal inserts based on the dominator tree but decided it wasn't worth it initially, but never revisisted if there's any benefit to inserting as close to the use as possible. Things are kept much simpler by always inserting in the first possible insertion point (impact described in forthcoming top-level PR comment).

asb commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,212 @@
+        InsertionPt = UserInst;
+      }
+
+      auto CacheKey = std::make_pair(Const, InsertionBB);
```
As noted above, I've removed this logic altogether for now as its quite subtle but doesn't have a big positive impact to justify it.

asb commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,212 @@
+    if (isa<PHINode>(I))
+      continue;
+    for (Use &U : I.operands()) {
+      if (auto *C = dyn_cast<ConstantFP>(U.get())) {
```
Just to confirm: Is this a concern about correctness or a suggestion for another profitable transform? Fully agreed it could be worthwhile doing promotion for vector constants and I'd like to return to that, but I'm trying to keep this very narrowly scoped initially.

arsenm commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,212 @@
+
+namespace {
+
+class RISCVPromoteConstant : public ModulePass {
```
Yes

preames replied:
On this particular suggestion, I would actively prefer this be done in a follow on change.  We do not have new pass manager versions of most RISCV codegen passes, and we should do these together.

preames commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,193 @@
+
+  for (Instruction &I : instructions(F)) {
+    for (Use &U : I.operands()) {
+      if (auto *C = dyn_cast<ConstantFP>(U.get())) {
```
Please use early-continue for this if

preames commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,193 @@
+        if (!C->getType()->isDoubleTy())
+          continue;
+        if (TLI->isFPImmLegal(C->getValueAPF(), MVT::f64,
+                              /*ForCodeSize=*/false))
```
This flag doesn't do anything on our target, but it made me realize we should probably be skipping this pass if optimizing for size.

preames reviewed (commented):
Close to an LGTM, but want to see the update and proof the tests one more time.

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q3-riscv-promote-constant-pass

asb committed:
Revert unneded test change and address review comments pt1

asb committed:
Address remaining review comments

preames commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,192 @@
+        InsertionBB = PN->getIncomingBlock(*U);
+      else
+        InsertionBB = UserInst->getParent();
+
```
I believe the current logic here is unsound when the predecessor contains a catchswitch since there is no valid insert point.  I'd suggest simply bailing in that case.

preames commented on llvm/lib/Target/RISCV/RISCVPromoteConstant.cpp:
```diff
@@ -0,0 +1,192 @@
+
+      // It is always safe to insert in the first insertion point in the BB,
+      // so do that and let other passes reorder.
+      InsertionPt = InsertionBB->getFirstInsertionPt();
```
Looks like this is only used in the else block below.

preames reviewed (approved):
LGTM w/comments addressed.

preames commented:
Reverse ping?

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q3-riscv-promote-constant-pass

asb committed:
Merge remote-tracking branch 'origin/main' into 2025q3-riscv-promote-constant-pass

asb committed:
Don't promote constant doubles for immarg args to intrinsics

I don't believe we have any intrinsics with fp immarg that I can test
this with.

asb committed:
Add note about inline asm

asb committed:
Add missed include

asb committed:
Address review comments

asb committed:
Drop unwanted changes from irtranslater/vararg.ll
